src/vs/base/parts/tree/browser/treeModel.ts

```diff
@@ -968,9 +970,18 @@ export class TreeModel {
 
 		let eventData: IRefreshEvent = { item: item, recursive: recursive };
 		this._onRefresh.fire(eventData);
-		return item.refresh(recursive).then(() => {
-			this._onDidRefresh.fire(eventData);
-		});
+		await item.refresh(recursive);
+		/**
+		 * Refresh loading status for item will be instantly lost upon refresh of item self
```

Comment:
So am I correct in my understanding that this is how it currently works? (especially the 1sec delay time)

1. We add the loading trait to the item before starting the refresh
2. The item "finishes" the refresh call immediately so the loading spinner disappears (but the actual refresh call is still happening in the background because of how OE does refreshes)
3. You add the loading trait back in again because it's not actually "done". But this loading spinner is hardcoded to always be around for 1sec - no matter how long the refresh takes
4. Eventually it finishes the refresh and actually updates the node

Reply:
Yeah that is correct, it isn't accurate to actual loading status, but the addTrait call requires an item to update, and I'm not sure where in the code does the actual refresh finishes (I tried to dig through, but I could never find it while debugging as there were a ton of recursion)

Reply:
Ok, well that doesn't really sound like a good fix to me. It's going to pretty much always be wrong - and refreshes can take a long time so if they take >1sec then it's just going to be confusing ("the loading spinner went away but the nodes were the same!")

If we're going to fix this it should be done correctly. (keep in mind that other things like the connection browse tree in the connection widget use this same tree structure so you probably shouldn't be inserting any connection-specific logic into the tree)